### PR TITLE
Admin edit-user-profile feature

### PR DESCRIPTION
**Pull request by patmbolger**
PT Story: https://www.pivotaltracker.com/story/show/164929174

Changes proposed in this pull request:
1. Added AdminOnly user_profile_controller, profile edit form, associated policies, etc., which allows admin to edit a user's profile attributes: names, emails (inc. the application _contact_email_), membership_number and password
2. Added link (edit icon) to users index view - goes to the user profile edit form.
3. Added ability for the admin to "become a user" via a button in the profile edit form - this signs in the admin _as_ that user, without having to enter user credentials - useful for checking changes from the user POV, seeing a problem that the user (but not admin) sees, switching to a user in development, etc.
4. Removed ability to edit user's membership_number in the application form.
  a. Removed associated methods (that allowed that editing) from ShfApplication.
5. Removed ability to edit user's password in the user show view.

Screenshots (Optional):

### Edit icon in users index view:

edit icon in users index view

### Edit user-profile form (admin-only):

admin edit user profile

Ready for review:
@AgileVentures/shf-project-team

### CODE REVIEW

File: config/locales/en.yml [hunk 1]
```diff
@@ -1031,6 +1031,18 @@ en:
           tab-title: &payments_h_branding "Payments: H-Branding"
           title: *payments_h_branding
 
+    user_profile:
```
Comment (weedySeaDragon) [inline config/locales/en.yml:1037]: per Rails convention, this should go under the action.  so these should go under `users:` (`update:` already exsits there, so the entries should just be added to that section)

(Same with the `sv.yml` entries)

File: config/locales/sv.yml [hunk 2]
```diff
@@ -1032,6 +1032,16 @@ sv:
           tab-title: &payments_h_branding "Betalningar: H-Branding"
           title: *payments_h_branding
 
+    user_profile:
```
Comment (weedySeaDragon) [inline config/locales/sv.yml:1038]: These go under the `users:` entry.  (See comment for en.yml)

File: config/locales/en.yml [hunk 3]
```diff
@@ -1181,6 +1193,8 @@ en:
         new_password: New password
         info_header: Edit name and e-mail
         password_confirmation: *password_confirmation
+
+    shared:
```
Comment (weedySeaDragon) [inline config/locales/en.yml]: `member_photo:pic_header` should go under `devise:registrations:edit`
(The organization is about the action, not the partial file name)

File: config/locales/sv.yml [hunk 4]
```diff
@@ -1179,6 +1189,8 @@ sv:
         new_password: Nytt lösenord
         info_header: Redigera namn och e-post
         password_confirmation: *password_confirmation
+
+    shared:
```
Comment (weedySeaDragon) [inline config/locales/sv.yml]: `member_photo:pic_header` should go under `devise:registrations:edit`
(The organization is about the action, not the partial file name)

File: config/routes.rb [hunk 5]
```diff
@@ -38,6 +38,9 @@
     get 'app_configuration/redigera', to: 'app_configuration#edit',
         as: :edit_app_configuration
 
+    get 'user_profile_edit/:id', to: 'user_profile#edit', as: :user_profile_edit
```
Comment (weedySeaDragon) [inline config/routes.rb:41]: I know we're not entirely consistent with the route naming, but in general we have used Swedish words.  So using Swedish and following the standard Rails routing patterns (and how we're generally doing the paths), these should be:
```
get 'anvandare/:id/redigera, to: 'user_profile#edit', as: :user_profile_edit
put 'anvandare/:id', to: 'user_profile#update', as: :user_profile_update
get 'anvandare/:id/<swedish word for "become">', to: 'user_profile#become', as: :become_user
```

File: spec/policies/admin_only_policy_spec.rb [hunk 6]
```diff
@@ -9,7 +9,7 @@
 
   let(:simple_record) { create(:business_category) }
 
-  CRUD_ACTIONS = [:index, :show, :new, :create, :edit, :update, :destroy]
+  CRUD_ACTIONS = [:index, :show, :new, :create, :edit, :update, :destroy, :become]
```
Comment (weedySeaDragon) [inline spec/policies/admin_only_policy_spec.rb:12]: "become" is not a CRUD action.  It doesn't belong in this array.

File: spec/policies/admin_only_user_profile_policy_spec.rb [hunk 7]
```diff
@@ -0,0 +1,48 @@
+require 'rails_helper'
+
+RSpec.describe AdminOnly::UserProfilePolicy do
+
+  let(:user_1) { create(:user, email: 'user@random.com') }
+  let(:member) { create(:member_with_membership_app, email: 'member@random.com', company_number: '5562728336')}
+  let(:admin)  { create(:user, email: 'admin@shf.com', admin: true) }
+  let(:visitor) { build(:visitor) }
+
+  CRUD_ACTIONS = [:edit, :update, :become]
```
Comment (weedySeaDragon) [inline spec/policies/admin_only_user_profile_policy_spec.rb]: "become" is not  a CRUD action.  doesn't belong here.

File: spec/policies/admin_only_user_profile_policy_spec.rb [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+
+  describe 'Admin is permitted everything' do
+
+    subject { described_class.new(admin, nil) }
+
+    CRUD_ACTIONS.each do | action |
+      it { is_expected.to permit_action action }
+    end
+  end
```
Comment (weedySeaDragon) [inline spec/policies/admin_only_user_profile_policy_spec.rb:19]: Should add another test here for `:become`; it's a test _in addition to_ the CRUD actions.

File: spec/policies/admin_only_user_profile_policy_spec.rb [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+
+
+  describe 'Member is forbidden everything' do
+    subject { described_class.new(member, nil) }
+
+    CRUD_ACTIONS.each do | action |
+      it { is_expected.to forbid_action action }
+    end
```
Comment (weedySeaDragon) [inline spec/policies/admin_only_user_profile_policy_spec.rb:27]: Should add another test here for `:become`; it's a test _in addition to_ the CRUD actions.

File: spec/policies/admin_only_user_profile_policy_spec.rb [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+  end
+
+
+  describe 'User (logged in) is forbidden everything' do
+    subject { described_class.new(user_1, nil) }
+
+    CRUD_ACTIONS.each do | action |
+      it { is_expected.to forbid_action action }
+    end
+  end
```
Comment (weedySeaDragon) [inline spec/policies/admin_only_user_profile_policy_spec.rb:37]: Should add another test here for `:become`; it's a test _in addition to_ the CRUD actions.

File: spec/policies/admin_only_user_profile_policy_spec.rb [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+
+  describe 'Visitor (not logged in) is forbidden everything' do
+    subject { described_class.new(visitor, nil) }
+
+    CRUD_ACTIONS.each do | action |
+      it { is_expected.to forbid_action action }
+    end
+  end
```
Comment (weedySeaDragon) [inline spec/policies/admin_only_user_profile_policy_spec.rb:45]: Should add another test here for `:become`; it's a test _in addition to_ the CRUD actions.

**Review by weedySeaDragon — CHANGES_REQUESTED** [review]
Comments about stuff _other than_ what the controller does to "become" a user.

File: config/locales/en.yml (hunk 1 above)
Comment (patmbolger) [reply config/locales/en.yml:1037]: I don't think the translations can be found that way.  I'm using the lazy lookup for controllers and views: https://guides.rubyonrails.org/i18n.html#lazy-lookup

Comment (patmbolger) [reply config/locales/en.yml:1037]: The controller is `user_profile`, and the actions are listed under that.

File: config/locales/en.yml (hunk 3 above)
Comment (patmbolger) [reply config/locales/en.yml]: Done.

File: config/routes.rb (hunk 5 above)
Comment (patmbolger) [reply config/routes.rb:41]: All of the "admin" routes seem to be have English URL's.  The user-visible routes seem to all(?) in Swedish, which is good.

I'd like to propose we make this kind of change through another story - if we decide that we need to adhere to the same standard for admin actions (which a user/member should never see).

File: spec/policies/admin_only_user_profile_policy_spec.rb (hunk 7 above)
Comment (patmbolger) [reply spec/policies/admin_only_user_profile_policy_spec.rb]: Done.

File: config/routes.rb (hunk 5 above)
Comment (thesuss) [reply config/routes.rb:41]: Agree to have the route-stuff in another story. One or more. :)

Comment (weedySeaDragon) [reply config/routes.rb:41]: agreed - another story is fine.

File: config/locales/en.yml (hunk 1 above)
Comment (weedySeaDragon) [reply config/locales/en.yml:1037]: Hm.  I've been using lazy lookup too.  If it works with how you have them organized then you way is right.  (And that teaches me something more about the organization.)

**Review by weedySeaDragon — APPROVED** [review]

**Pull request merged**